### PR TITLE
Better rholang parallelism

### PR DESCRIPTION
**Pull request by ArturGajowy**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

Benchmarks in the updated [spreadsheet](https://docs.google.com/spreadsheets/d/11WdCneacNV4yWD14TXz2WKL-PxGYQbFGiAiXPNF0V-8/edit#gid=0)

The log messages for benchmarks were output using the code on [this branch](https://github.com/ArturGajowy/rchain/commits/better_rholang_parallelism_extras)

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by ArturGajowy** [comment]
Bors try

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 1]
```diff
@@ -178,26 +179,62 @@ class DebruijnInterpreter[M[_], F[_]](
...
+      // - the scheduler used
+      // - the monix ExecutionModel used
+      // Verify if using traverse-per-batch inside parTraverse means that less tasks are spawned at once
+      // (surmise: the task for the next element of a batch is not spawned until the previous one is processed)
+      // Track memory usage along with raw performance when investigating. Clean slate POC could be helpful here.
+
+      // we index terms before shuffling to have deterministic Blake2b512Random seeds for each term
+      val indexedTermsShuffled = Random.shuffle(indexedTerms)
+      val groups               = indexedTermsShuffled.groupBy(_._2 % Reduce.parallelism).values
+      groups.toList.parTraverse_ {
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: @pyrocto and others - do you see an issue in here?

Notice we're retaining the random seed by indexing the terms prior to shuffling.

I think this is a faithful, if not slightly idealized, simulation of parallel execution order of pars. Surely it's still non-deterministic, probably even more than the previous approach. It should also bring more fairness and bugs found due to a more uniform distribution of execution orders too.

Does anybody see issues with the approach?

**Comment by ArturGajowy** [comment]
bors try

**Comment by JosephDenman** [comment]
This is a nice optimization. Obviously, the claim that it is an optimization should be verified before the PR is merged, but it looks promising.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 1 above)
Comment (ghost) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: This looks fine to me.

**Review by ghost — APPROVED** [review]

**Comment by ArturGajowy** [comment]
> AFAIK the order of the indices doesn’t matter.

It does matter. Each par needs to have its own, deterministc random space. We achive that by sorting the pars in the canonical order and doing `.split(index)` on the parent BlakeRandom

**Comment by ArturGajowy** [comment]
> Obviously, the claim that it is an optimization should be verified before the PR is merged

Have you seen the updated to the spreadsheet? (link above)
If you need more, feel free to run your own benchmarks.

**Comment by JosephDenman** [comment]
> It does matter. Each par needs to have its own, deterministc random space. We achive that by sorting the pars in the canonical order and doing .split(index) on the parent BlakeRandom

I know why we need to split BlakeRandom. I’m saying it doesn’t matter which index we use to split it. So, there’s no need to retain the original index when shuffling.

**Comment by ArturGajowy** [comment]
>  it doesn’t matter which index we use to split it. 

How do you replay then?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 2]
```diff
@@ -178,26 +179,46 @@ class DebruijnInterpreter[M[_], F[_]](
...
-              case e: EVarBody    => reportErrors(eval(e.value.v) >>= (eval(_)))
-              case e: EMethodBody => reportErrors(evalExprToPar(e) >>= (eval(_)))
-              case other          => BugFoundError(s"Undefined term: \n $other").raiseError[M, Unit]
-            }
-          case other => BugFoundError(s"Undefined term: \n $other").raiseError[M, Unit]
-        }
+    // we index terms before shuffling to have deterministic Blake2b512Random seeds for each term
+    val indexedTermsShuffled = Random.shuffle(terms.zipWithIndex)
+    val groups               = indexedTermsShuffled.groupBy(_._2 % Reduce.parallelism).values
+    groups.toList.parTraverse_ {
```
Comment (dzajkowski) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: nit: `toList` can go on the prev line as it's noise

File: rspace/src/main/scala/coop/rchain/rspace/history/HistoryRepositoryImpl.scala [hunk 3]
```diff
@@ -193,7 +193,7 @@ final case class HistoryRepositoryImpl[F[_]: Sync, C, P, A, K](
 
   override def checkpoint(actions: List[HotStoreAction]): F[HistoryRepository[F, C, P, A, K]] =
     for {
-      trieActions    <- Applicative[F].pure(transform(actions))
+      trieActions    <- Applicative[F].pure(actions.par.map(transform).toList)
```
Comment (dzajkowski) [inline rspace/src/main/scala/coop/rchain/rspace/history/HistoryRepositoryImpl.scala:196]: note: I will be revamping this whole computation in my next pr

**Review by dzajkowski — APPROVED** [review]

**Pull request merged**